File: examples/02-reliability/java/bounded_fanout.java
// Real-world: fan out 10k notifications, but a downstream provider only
// tolerates ~20 in-flight calls. Unbounded Async.function would schedule all
// 10k at once. A deterministic in-Workflow counter gated by Workflow.await caps
// concurrency while keeping the pipeline full — the durable equivalent of a
// bounded thread pool. (Don't use java.util.concurrent.Semaphore here: blocking
// a real thread isn't replay-safe. Workflow.await is the deterministic wait.)
class BoundedFanoutWorkflow {
  private final NotificationActivities notify =
      Workflow.newActivityStub(
          NotificationActivities.class,
          ActivityOptions.newBuilder().setStartToCloseTimeout(Duration.ofMinutes(1)).build());

  int sendAll(List<String> userIds, int maxInFlight) {
    int[] inFlight = {0};
    List<Promise<Void>> sends = new ArrayList<>();

    for (String id : userIds) {
      // Park the Workflow (deterministically) until a slot frees up.
      Workflow.await(() -> inFlight[0] < maxInFlight);
      inFlight[0]++;
      Promise<Void> send =
          Async.procedure(notify::send, id)
              .thenApply(ignored -> { inFlight[0]--; return null; });
      sends.add(send);
    }

    Promise.allOf(sends).get();
    return userIds.size();
  }
}

File: examples/02-reliability/go/bounded_fanout.go
package reliability

import (
	"time"

	"go.temporal.io/sdk/workflow"
)

// Real-world: fan out 10k notifications, but a downstream provider only
// tolerates ~20 in-flight calls. A deterministic in-Workflow counter gated by
// workflow.Await caps concurrency while keeping the pipeline full — the durable
// equivalent of a bounded worker pool. (Don't use a Go sync primitive here;
// workflow.Await is the replay-safe wait.)
func SendAll(ctx workflow.Context, userIDs []string, maxInFlight int) (int, error) {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: time.Minute,
	})

	inFlight := 0
	futures := make([]workflow.Future, 0, len(userIDs))
	for _, id := range userIDs {
		// Park the Workflow (deterministically) until a slot frees up.
		_ = workflow.Await(ctx, func() bool { return inFlight < maxInFlight })
		inFlight++
		f := workflow.ExecuteActivity(ctx, "Send", id)
		workflow.Go(ctx, func(ctx workflow.Context) {
			_ = f.Get(ctx, nil)
			inFlight--
		})
		futures = append(futures, f)
	}

	for _, f := range futures {
		_ = f.Get(ctx, nil)
	}
	return len(userIDs), nil
}
